src/runtime/hexagon_remote/halide_remote.cpp

```diff
@@ -44,16 +87,227 @@ int halide_do_task(void *user_context, halide_task_t f, int idx,
+        if (!owned_job) {
+            // If this threads doesn't own a job, it looks for one
+            // and tries to do it. If it cannot find a job, it goes
+            // to sleep.
+            job = work_queue.jobs;
+            if (!job) {
+                // release the lock and go to sleep.
+                // ***************************
+                // *** Work queue unlocked ***
+                // ***************************
+                FARF(LOW, "HVX_TP: %d: goto_work: Couldn't find a job, going to sleep\n", tid);
+                unlock_work_queue();
+                qurt_sem_down(&wait_for_work);
+                FARF(LOW, "HVX_TP: %d: goto_work: Got woken up. Going to look for work.\n", tid);
+                 // ***********************
+                // *** Lock work queue ***
+                // ***********************
+                lock_work_queue();
+                continue;
+            } else {
+                myjob = *job;
+                FARF(LOW, "HVX_TP: %d: goto_work: Found a job working on x = %d\n", tid, myjob.next);
+            }
+        } else {
+            // We are here only if this thread owns a job that is not done.
+            // So do a part of it.
+            if (owned_job->claimed()) {
+                // the owner goes to sleep till the workers wake it up.
+                FARF(LOW, "HVX_TP: %d: goto_work: Owner about to sleep\n", tid);
+                qurt_cond_wait(&owned_job->wakeup_owner, &work_queue.work_mutex);
+                FARF(LOW, "HVX_TP: %d: goto_work: Owner waking up.\n", tid);
+                // This thread should have the lock now after having been woken up.
+                break;
+            } else {
+                myjob = *owned_job;
+                job = owned_job;
+                FARF(LOW, "HVX_TP: %d: goto_work: Owner about to work\n", tid);
+            }
+        }
+        job->next++;
+        job->active_workers++;
+
+        // If all tasks of the job have been claimed, then pop the job off the stack.
+        if (job->claimed()) {
+            work_queue.jobs = job->next_job;
+            FARF(LOW, "HVX_TP: %d: goto_work: All tasks in job claimed. Popping job from stack.\n", tid);
+        }
+        // ***************************
+        // *** Work queue unlocked ***
+        // ***************************
+        unlock_work_queue();
+
+        FARF(LOW, "HVX_TP: %d: goto_work: About to do_task, user_context = 0x%x, f = 0x%x x = %d \n", tid, myjob.user_context, myjob.f, myjob.next);
+        int result = halide_do_task(myjob.user_context, myjob.f,
+                                    myjob.next, myjob.closure);
+        FARF(LOW, "HVX_TP: %d: goto_work: Finished do_task with status = %d\n", tid, result);
+        // ***********************
+        // *** Lock work queue ***
+        // ***********************
+        lock_work_queue();
         if (result) {
-            return result;
+            job->exit_status = result;
+        }
```

Comment:
One way we use this is to cancel pipelines mid-processing, in which case the only reason to support this is to return from the parallel work quickly.

Looking at it more closely, returning from the thread isn't the right thing to do, but we should just stop executing new work if the exit_status is non-zero, so the thread pool will burn through the new work very quickly and exit the thread pool normally.

Maybe set job->max = job->next, in addition to exit_status? That way the thread pool will not start any new work after an error.